feat: sound effects from protocol 13x

marcosvf132 opened the pull request:
Client: https://github.com/dudantas/tibia-client/pull/6

# Preview
https://user-images.githubusercontent.com/66353315/193288117-be5d4df1-d6ac-4a07-8e6d-84cf00542327.mp4

# 13.10
![image](https://user-images.githubusercontent.com/8551443/208708297-63810753-eb63-4f09-9e73-71856bad6fb4.png)

# WIP
- This PR implement a huge amount of features ingame that need time and test to verify all of then and to adjust the details on it.

# Musics and anthem
- Musics and anthem are not on this PR yet because i'm still catalogating then all, check for further commits on this PR.

# Features implemented so far

## Player combat
- Player weapon sounds are implemented, with future adjustments like a specifc sound on a different item.
- Wand and rods so far have the same sound, which will be changed on the next commits.

## Monster combat
- There is a script that run on monster spell declarations that automaticlly detect which spell sound it should be.
- If the user want's to select a custom sound to a spell, can use the following tags:
![image](https://user-images.githubusercontent.com/66353315/193341050-79fd423c-ebcb-4ae3-9c66-4071d66216b3.png)
![image](https://user-images.githubusercontent.com/66353315/193341073-44f95b87-b5bc-459e-87c7-e8252dcd7500.png)


## Item movement
- All items have sound on their movement, some can have the wrong sound, if you know an item that does not have the sound so accurated, please help me improving it.

## Npc sounds
- The NPC Canary does emit sounds on his sourrounds, check the new functions label on this description for more information.

# New function

# Enums
- We have  a LOT of sounds on tibia (512 so far), all those sounds are catalogated on the **/src/creatures/creatures_definitions.hpp** file, rigth here:
![image](https://user-images.githubusercontent.com/66353315/193290273-08f5ef60-5199-4fa7-b184-5aea27b6a77d.png)
###       NOTE: You can use then all on LUA env


# C++

## Sound movement
- The item sound movement comes from this new function, based on item loot category. I know this is not 100% accurated but its the best shot so far.
- The sound can change if the it's is being moved inside depot or player backpacks/inventory on the game.cpp call.
![image](https://user-images.githubusercontent.com/66353315/193290783-600baa7f-3e66-4992-9d1b-6e7fef4a3e17.png)

## Sound notification to spectators
- On this topic we have 2 new functions, the first one is used to send a unique sound to the spectators on the area. It already check the sound source if it's coming from the own player, other player, a creature or it's meant to be a global sound. This source type is used on the client to balance and mix the volume as the player desire.
![image](https://user-images.githubusercontent.com/66353315/193291336-5b8b305a-9b26-465d-a48e-96d0e326c3c3.png)
- This second function is used on distance weapons or any other sound that must be reproduced on a sequence. Like a casting sound and an impact sound.
![image](https://user-images.githubusercontent.com/66353315/193291575-ff582159-1489-4cca-8306-198ed5fbe728.png)

## OnThink sounds
- Some creatures emits sounds while in battle or when you get close to then. The first function is for the NPCs:
![image](https://user-images.githubusercontent.com/66353315/193292147-123c16df-c2af-4364-95f8-5944263a4c8a.png)
- The monsters sounds:
![image](https://user-images.githubusercontent.com/66353315/193292296-3c6eb5bb-1a25-410a-90d1-a37d5239a56f.png)

## Player client functions
- The player.h functions are used to send the sound to the client, can be used to send a sound directly to the player client and not send to all the players on the area.
![image](https://user-images.githubusercontent.com/66353315/193293086-516315f5-65a2-49ad-b299-5b01e186fe29.png)

## Player sound functions
- Functions that return the sound ID by the weapond attack and impact sound.
- Impact function:
![image](https://user-images.githubusercontent.com/66353315/193293341-70431b29-e682-4971-999b-eaf8ca23fdd3.png)
- Attack function:
![image](https://user-images.githubusercontent.com/66353315/193293410-e654590b-6562-4d19-8705-fb8a3e6b3505.png)

# LUA

## Spell (local spell = Spell("spell"))
- We have the possibility to add sounds on the spell, can be the sound when casting the spell AND the sound when the spell hits the target/area.
- Casting sound:
![image](https://user-images.githubusercontent.com/66353315/193293797-008812fb-dda3-47cf-90de-151ef984fc22.png)
- Impact sound:
![image](https://user-images.githubusercontent.com/66353315/193293854-9b222cd4-e237-4e63-923d-c6479534b684.png)

## Bestiary charm
- We have the possibility to add sounds on the spell, can be the sound when casting the spell AND the sound when the spell hits the target/area.
![image](https://user-images.githubusercontent.com/66353315/193294228-83387fa1-6ef4-4357-a3c7-4284ca550ee4.png)

## Npc
- You can add sound effects on the NPC just like the old text system, can be used as a environment sound as a blacksmith sound, hitting a wood, casting a spell etc.
- A random sound will be picked on the list to be reproduced.
![image](https://user-images.githubusercontent.com/66353315/193294768-4eaa436a-b8d2-4802-8803-5ac71a8347c6.png)

## Position
- You can used this function to send sounds on the specific position, just like the position:sendMagicEffect(...)
![image](https://user-images.githubusercontent.com/66353315/193295067-4964bb5b-ee21-4667-920d-909fb0f945f7.png)
![image](https://user-images.githubusercontent.com/66353315/193295095-5d706feb-15b9-43bb-984c-a2cd7d279ae0.png)
### The optional 'actor' on this function right here can be used to specify if you want to send the sound pretending is was coming from the creature. If you don't send a creature on the param, it will be sent as a global sound.

# XML
- The imbuement system can emits sounds too, you can set a sound on the imbuements.xml files be using the id:
![image](https://user-images.githubusercontent.com/66353315/193295564-6098dba7-f309-4ad2-be44-b3255c621009.png)

marcosvf132 committed:
Initial commit

marcosvf132 committed:
Merge branch 'main' into test

marcosvf132 committed:
Automaticlly detect monster spells sonds + dat update

- Remove print from charm sounds
- Fix condition attr bug

marcosvf132 committed:
Merge branch 'test' of https://github.com/Multiple-Forks/optimized-canary into test

marcosvf132 committed:
Player spells

MatheusAugustoGarcia commented on src/lua/functions/creatures/player/player_functions.hpp:
```diff
@@ -275,7 +275,8 @@ class PlayerFunctions final : LuaScriptInterface {
 
 			registerMethod(L, "Player", "isOffline", PlayerFunctions::luaPlayerIsOffline);
 
-			registerMethod(L, "Player", "openMarket", PlayerFunctions::luaPlayerOpenMarket);
```
Apagou sem querer?

marcosvf132 committed:
Combat changes declaration (1/2)

marcosvf132 committed:
Combat changes declaration (2/2)

marcosvf132 committed:
Update appearances.dat (13.05)

marcosvf132 committed:
Merge branch 'main' into test

marcosvf132 committed:
Store fix

Peixonauta2 commented:
It seems that after the last changes, spells and runes are not playing any sounds. Was it tested after these few changes?

I've noticed that the limits to max health and max mana were changed, but the mana display is still locked at 65535. Is this intended?

For the rest, it looks great!

marcosvf132 committed:
Fix spells sounds + mana limit on 65k

Peixonauta2 commented:
Nice! That fixed it.

With that, I've noticed that monsters are not playing sounds when casting a script spell.

I've added this code to `bool InstantSpell::executeCastSpell(Creature* creature, const LuaVariant& var)` and it fixed it temporarily
`	if (soundCastEffect != SOUND_EFFECT_TYPE_SILENCE) {`
`		g_game().sendDoubleSoundEffect(creature->getPosition(), soundCastEffect, soundImpactEffect, creature);`
`	}`


Still don't know if that is the best choice.

Peixonauta2 commented:
It seems that there's a problem with this new way to handle the store. After some interaction with the menu, the client crashes:

![Animação](https://user-images.githubusercontent.com/63183357/197430062-2e0fd691-0b3a-4c8e-8008-083cd24b2837.gif)

After that crash, every time **any** character tries to open the store, the client will crash. Server restart is required to "fix" it until someone make it crashable again.

After buying a Hireling, the client also crashes.

tutbarao commented:
@Peixonauta2 Did you manage to solve it?

aokisama commented:
Same issue, error seems to be somewhere here in the disabled reasons but unsure on how to fix it.

```
local canBuy = player:canBuyOffer(offer)
		if (canBuy.disabled == 1) then
			for index, disableTable in ipairs(disableReasons) do
				if (canBuy.disabledReason == disableTable.reason) then
					offer.disabledReadonIndex = index
				end
			end

			if (offer.disabledReadonIndex == nil) then
				offer.disabledReadonIndex = #disableReasons
				table.insert(disableReasons, canBuy.disabledReason)
			end
		end
```

dudantas committed:
Merge branch 'main' into pr/531

dudantas committed:
Sound lua enums

dudantas committed:
Merge branch 'main' into test

dudantas committed:
Merge branch 'main' into pr/531

dudantas committed:
Support to 13.10

Peixonauta2 commented:
As far as I've tested, the problems I've found are in the spells used by monsters that still don't play any sound - I guess that there is missing the new function to play a sound in InstantSpell::executeCastSpell,  and the problem of the store crashing the client after a few interactions.

marcosvf132 committed:
Update init.lua

dudantas committed:
Some fixes

dudantas committed:
Merge branch 'main' into test

dudantas commented:
> As far as I've tested, the problems I've found are in the spells used by monsters that still don't play music - I guess that there is missing the new function to play a sound in InstantSpell::executeCastSpell, and the problem of the store crashing the cliente after a few interactions.

Not the entire sound system will be done on this pull. Only a few things were implemented. As for the store, it has already been corrected.

dudantas committed:
Remove logs

dudantas committed:
Add sound to player spells

dudantas committed:
Fixing monster spells and others things

Peixonauta2 commented:
Everything seems fine by now.

dudantas committed:
Merge remote-tracking branch 'upstream/main' into pr/531

dudantas committed:
Add sound methods to player functions

dudantas committed:
Conflict fixes

dudantas committed:
Compile fix and change to 13.11

dudantas committed:
Some sonar check fixes

dudantas committed:
Sonar fixes and others things

Added new functions to tools.h:
The first function, "convertToSafeInteger", is a template function that takes in two types: "ReturnValue" and "ConvertValue", as well as a value of "convertValue" of type "ConvertValue". It converts this value to the "ReturnValue" type and ensures that it falls within the range of 0 and the maximum value that the "ReturnValue" type can hold. It then returns the converted and safe value.

The second function, "getEnumClassNumber", is also a template function that takes in a single type: "EnumClass", and a reference to a value of "convertValue" of type "EnumClass". It uses the "magic_enum" library to convert the enumeration value to its underlying integral value and returns it.

dudantas committed:
Fix conversion

dudantas committed:
Appearances 13.11

dudantas committed:
Fixes compilation

dudantas committed:
Sonar fixes

dudantas committed:
More sonar fixes

dudantas committed:
Merge remote-tracking branch 'upstream/main' into pr/531

dudantas committed:
Sonar fixes

dudantas commented:
> Everything seems fine by now.

Can you test again? We've made several changes related to wrong conversions. It will probably be necessary to test several things, from attack, combat, mana, health, etc.

Olimpotibia commented:
Tested Yesterday in a fresh machine and runs without any issue & stable.

saitamOsirb commented:
I was testing this code and it causes a server crash when using objects or food

saitamOsirb commented:
When selecting an object in the market and pressing the down key and keeping it pressed for a while, the tibia client breaks

saitamOsirb commented:
When attacking a creature, it stays red alive on first hit and doesn't die.

saitamOsirb commented:
when walking with the character outside the city the character walks without movement it is as if he were floating

dudantas committed:
Merge remote-tracking branch 'upstream/main' into pr/531

dudantas committed:
Update protocolgame.cpp

a

github-actions[bot] committed:
Code format - (Clang-format)

dudantas commented:
We are going to develop this branch again, soon we will have news about the bug reports that have been made.

dudantas commented:
> I was testing this code and it causes a server crash when using objects or food

![image](https://user-images.githubusercontent.com/8551443/218583633-288626ce-6110-4d7a-84e2-82ead6464458.png)
![image](https://user-images.githubusercontent.com/8551443/218583673-61e9772d-dc2a-4e5f-a9b6-b42853485de1.png)
![image](https://user-images.githubusercontent.com/8551443/218583685-5748c629-d31e-45de-9f64-de55a7caa432.png)

You need to be more specific... Which item? I was unable to reproduce crash.

saitamOsirb commented:
It was fixed in a commit that said fix condition a few days ago that was uploaded to this pr. Only the bugs that the characters walk badly outside the city or when you take out skull and the bugs with the life charge of the creatures when you hit them that they remain black having a lifetime

dudantas commented:
> It was fixed in a commit that said fix condition a few days ago that was uploaded to this pr. Only the bugs that the characters walk badly outside the city or when you take out skull and the bugs with the life charge of the creatures when you hit them that they remain black having a lifetime

Ok... I'm looking into this one right now.

dudantas committed:
fix: unique loot

dudantas committed:
fix: creatures health

dudantas committed:
Merge branch 'test' of https://github.com/Multiple-Forks/optimized-canary into pr/531

dudantas commented:
> When attacking a creature, it stays red alive on first hit and doesn't die.

Fixed. Thanks for reporting.

dudantas committed:
fix some things and add imbuement tracker

github-actions[bot] committed:
Code format - (Clang-format)

dudantas committed:
Sanity check

dudantas committed:
Merge remote-tracking branch 'upstream/main' into pr/531

dudantas committed:
fix: imbuement decay tracker

dudantas committed:
Revert

dudantas committed:
Fix some bugs

dudantas committed:
fix: depot items not saving

elsongabriel commented on src/io/iologindata.cpp:
```diff
@@ -505,6 +505,36 @@ bool IOLoginData::loadPlayer(Player* player, DBResult_ptr result) {
 
 	// load depot items
 	itemMap.clear();
+	query.str(std::string());
```
it's already added below.

elsongabriel commented on src/io/iologindata.cpp:
```diff
@@ -505,6 +505,36 @@ bool IOLoginData::loadPlayer(Player* player, DBResult_ptr result) {
+				}
+			}
+		}
+	}
```
to here.. in line (545)

4190 commented:
> > When attacking a creature, it stays red alive on first hit and doesn't die.
> 
> Fixed. Thanks for reporting.

There's similar bug now. When you score a bloodhit on something health status does not update- game shows it to be full hp no matter how much hp creature has.

dudantas committed:
Merge remote-tracking branch 'upstream/main' into pr/531

dudantas committed:
fix conflict

dudantas committed:
fix health

dudantas committed:
fix function convertToSafeInteger logic

first it will try to convert using the static cast, only if it does not succeed, it will use the other forms of conversion.

dudantas committed:
fix

dudantas committed:
fix conflict

4190 commented:
Mixed up display of club fighting and distance fighting in client.

When you skill up dist skill- bar under and number next to club fighting goes up 
Same when you skill up club skill - bar under and number next to distance fighting goes up

Skill itself works correctly it just got mixed up in display.

![obraz](https://user-images.githubusercontent.com/28237211/221065554-b0326a4a-412e-4d86-9eac-1f2e91b51cb5.png)

dudantas committed:
fix: decimal skill to life and mana leech

dudantas committed:
Merge branch 'main' into pr/531

github-actions[bot] committed:
fix weapons sound and lua enums

Olimpotibia commented:
looks like sweet but server is some lagged sometimes.

dudantas committed:
improve: move lua environment init to otserv.cpp

dudantas committed:
fix: lua sound enums

github-actions[bot] committed:
Code format - (Clang-format)

4190 commented:
Cloning items bug. How to reproduce- 
Have some items in depot box (doesn't matter which one, let's use depot box number 1)
 
Put one of items on market for sale
Remove the sell offer so the item goes back to inbox
Logout and login
All of items from depot box get cloned into inbox.

dudantas committed:
fix: safe conversion of numbers using gsl library

github-actions[bot] committed:
Code format - (Clang-format)

dudantas committed:
fix type size

dudantas committed:
fix: mastermind potion bug and other errors

codigostrueno commented:
manaleach and lifeleach are displayed incorrectly from the xml to the client and also add much more than it should give in manaleach and lifeleach, this could be happening with all the imbuiments

codigostrueno commented:
when you have a shared party and a member dies, they are not expelled and the shared party is lost

codigostrueno commented:
when summoning a familiar and it dies it continues to appear infinitely

dudantas committed:
fix: send the log only if the diff has a difference of 1 or more

dudantas committed:
improve log

lucassaraiva5 commented:
> Cloning items bug. How to reproduce- Have some items in depot box (doesn't matter which one, let's use depot box number 1)
> 
> Put one of items on market for sale Remove the sell offer so the item goes back to inbox Logout and login All of items from depot box get cloned into inbox.

Can you explain better, I could not replicate

codigostrueno commented:
video evidence for cloning market https://www.youtube.com/watch?v=q2WFxA2806c

codigostrueno commented:
making look at items that have attribute absorb percentage does not show the element well

codigostrueno commented:
shutdown of the server at a specific time even if the configuration is false for automatic shutdown

codigostrueno commented:
no open corpse with reward  ,  send message you not are owner   (solution :  https://github.com/opentibiabr/canary/pull/885/files ?)

dudantas committed:
Merge branch 'main' into pr/531

dudantas committed:
fix: familiars creation

codigostrueno commented:
The familiar's error is not in the creation, that was in the event that calls the familiar, since when it dies, the familiar continues to appear infinitely many times.

codigostrueno commented:
evidence boss message loot 
![image](https://user-images.githubusercontent.com/127452519/225695015-d2fd8b74-3f11-4e57-9469-c0666bfc0376.png)

dudantas committed:
fix: debug hireling

codigostrueno commented:
utana vid condition not finish

dudantas committed:
Merge branch 'main' into pr/531

codigostrueno commented:
exura sio pulls out skull

dudantas committed:
fix: imbuement tracker

dudantas committed:
fix: from break to continue

codigostrueno commented:
error at look items with attribute  absorbpercentenergy, absorbpercentfire they are exchanged

dudantas commented:
> no open corpse with reward , send message you not are owner (solution : https://github.com/opentibiabr/canary/pull/885/files ?)

![image](https://user-images.githubusercontent.com/8551443/228398781-013cf209-a81b-452d-b263-ea0ad04af6e6.png)

dudantas commented:
> error at look items with attribute absorbpercentenergy, absorbpercentfire they are exchanged

it's not a PR bug, it's a main bug... but I already found the bug <https://github.com/opentibiabr/canary/pull/961>

dudantas committed:
some small fixes

dudantas committed:
Merge branch 'main' into pr/531

dudantas committed:
fix: hireling debug

codigostrueno commented:
when surrounding a teleport with magicwall in front of the teleport with script one player and another player connects on top of it the server breaks

codigostrueno commented:
when change wepons one hand  by two hand  is require first send shield to backpack , not change automatically

codigostrueno commented:
when a player dies in party share , the corpse keeps gaining exp and destroys the party shared, i send evidence in other comment

codigostrueno commented:
when a player death with blank skull and shot a magic field rune  revive with blank skull

codigostrueno commented:
inbuiment proyection elemental they do not work

codigostrueno commented:
any modification in atributes manaleach or lifelich , elemental no refresh

codigostrueno commented:
They are not well defined in terms of damage, some hit too much others hardly hit

dudantas committed:
fix: remove white skull on death

dudantas committed:
fix conflict errors

dudantas committed:
Merge branch 'main' into pr/531
